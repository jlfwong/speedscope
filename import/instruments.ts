// This file contains methods to import data from OS X Instruments.app
// https://developer.apple.com/library/content/documentation/DeveloperTools/Conceptual/InstrumentsUserGuide/index.html

import {Profile, FrameInfo, ByteFormatter, TimeFormatter} from '../profile'

function parseTSV<T>(contents: string): T[] {
  const lines = contents.split('\n').map(l => l.split('\t'))

  const headerLine = lines.shift()
  if (!headerLine) return []

  const indexToField = new Map<number, string>()
  for (let i = 0; i < headerLine.length; i++) {
    indexToField.set(i, headerLine[i])
  }

  const ret: T[] = []
  for (let line of lines) {
    const row = {} as T
    for (let i = 0; i < line.length; i++) {
      ;(row as any)[indexToField.get(i)!] = line[i]
    }
    ret.push(row)
  }
  return ret
}

interface PastedTimeProfileRow {
  Weight?: string
  'Source Path'?: string
  'Symbol Name'?: string
}

interface PastedAllocationsProfileRow {
  'Bytes Used'?: string
  'Source Path'?: string
  'Symbol Name'?: string
}

interface FrameInfoWithWeight extends FrameInfo {
  endValue: number
}

function getWeight(deepCopyRow: any): number {
  if ('Bytes Used' in deepCopyRow) {
    const bytesUsedString = deepCopyRow['Bytes Used']
    const parts = /\s*(\d+(?:[.]\d+)?) (\w+)\s+(?:\d+(?:[.]\d+))%/.exec(bytesUsedString)
    if (!parts) return 0
    const value = parseInt(parts[1], 10)
    const units = parts[2]

    switch (units) {
      case 'Bytes':
        return value
      case 'KB':
        return 1024 * value
      case 'MB':
        return 1024 * 1024 * value
      case 'GB':
        return 1024 * 1024 * 1024 * value
    }
    throw new Error(`Unrecognized units ${units}`)
  }

  if ('Weight' in deepCopyRow) {
    const weightString = deepCopyRow['Weight']
    const parts = /\s*(\d+(?:[.]\d+)?) (\w+)\s+(?:\d+(?:[.]\d+))%/.exec(weightString)
    if (!parts) return 0
    const value = parseInt(parts[1], 10)
    const units = parts[2]

    switch (units) {
      case 'ms':
        return value
      case 's':
        return 1000 * value
      case 'min':
        return 1000 * value
    }
    throw new Error(`Unrecognized units ${units}`)
  }

  return -1
}

// Import from a deep copy made of a profile
export function importFromInstrumentsDeepCopy(contents: string): Profile {
  const profile = new Profile()
  const rows = parseTSV<PastedTimeProfileRow | PastedAllocationsProfileRow>(contents)

  const stack: FrameInfoWithWeight[] = []
  let cumulativeValue: number = 0

  for (let row of rows) {
    const symbolName = row['Symbol Name']
    if (!symbolName) continue
    const trimmedSymbolName = symbolName.trim()
    let stackDepth = symbolName.length - trimmedSymbolName.length

    if (stack.length - stackDepth < 0) {
      console.log(stack, symbolName)
      throw new Error('Invalid format')
    }

    let framesToLeave: FrameInfoWithWeight[] = []

    while (stackDepth < stack.length) {
      const stackTop = stack.pop()!
      framesToLeave.push(stackTop)
    }

    for (let frameToLeave of framesToLeave) {
      cumulativeValue = Math.max(cumulativeValue, frameToLeave.endValue)
      profile.leaveFrame(frameToLeave, cumulativeValue)
    }

    const newFrameInfo: FrameInfoWithWeight = {
      key: `${row['Source Path'] || ''}:${trimmedSymbolName}`,
      name: trimmedSymbolName,
      file: row['Source Path'],
      endValue: cumulativeValue + getWeight(row),
    }

    profile.enterFrame(newFrameInfo, cumulativeValue)
    stack.push(newFrameInfo)
  }

  while (stack.length > 0) {
    const frameToLeave = stack.pop()!
    cumulativeValue = Math.max(cumulativeValue, frameToLeave.endValue)
    profile.leaveFrame(frameToLeave, cumulativeValue)
  }

  if ('Bytes Used' in rows[0]) {
    profile.setValueFormatter(new ByteFormatter())
  } else if ('Weight' in rows[0]) {
    profile.setValueFormatter(new TimeFormatter('milliseconds'))
  }

  return profile
}
