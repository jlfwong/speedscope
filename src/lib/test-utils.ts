import * as fs from 'fs'
import * as path from 'path'
import {Profile, CallTreeNode, Frame} from './profile'
import {importProfileGroup} from '../import'
import {exportProfileGroup, importSpeedscopeProfiles} from './file-format'

interface DumpedProfile {
  name: string
  stacks: string[]
  frames: Frame[]
}

export function dumpProfile(profile: Profile): any {
  const dump: DumpedProfile = {
    name: profile.getName(),
    stacks: [],
    frames: [],
  }

  profile.forEachFrame(f => dump.frames.push(f))

  let lastValue = 0
  const curStack: (number | string)[] = []

  function maybeEmit(value: number) {
    if (lastValue != value) {
      dump.stacks.push(
        curStack.map(k => `${k}`).join(';') + ` ${profile.formatValue(value - lastValue)}`,
      )
      lastValue = value
    }
  }

  function openFrame(node: CallTreeNode, value: number) {
    maybeEmit(value)
    curStack.push(node.frame.name)
  }

  function closeFrame(node: CallTreeNode, value: number) {
    maybeEmit(value)
    curStack.pop()
  }

  profile.forEachCall(openFrame, closeFrame)

  return dump
}

export async function checkProfileSnapshot(filepath: string) {
  const input = fs.readFileSync(filepath, 'utf8')

  const profileGroup = await importProfileGroup(path.basename(filepath), input)
  if (profileGroup) {
    expect(profileGroup.name).toMatchSnapshot('profileGroup.name')
    expect(profileGroup.indexToView).toMatchSnapshot('indexToView')

    for (let profile of profileGroup.profiles) {
      expect(dumpProfile(profile)).toMatchSnapshot()
    }
  } else {
    fail('Failed to extract profile')
    return
  }

  const profilesWithoutFilename = await importProfileGroup('unknown', input)
  if (profilesWithoutFilename) {
    expect(profilesWithoutFilename.profiles.length).toEqual(profileGroup.profiles.length)
    profilesWithoutFilename.name = profileGroup.name
    for (let i = 0; i < profileGroup.profiles.length; i++) {
      const a = profileGroup.profiles[i]
      const b = profilesWithoutFilename.profiles[i]
      b.setName(a.getName())
    }
    expect(exportProfileGroup(profileGroup)).toEqual(exportProfileGroup(profilesWithoutFilename))
  } else {
    fail('Failed to extract profile when filename was "unknown"')
    return
  }

  const exported = exportProfileGroup(profileGroup)
  const reimportedGroup = importSpeedscopeProfiles(exported)

  expect(reimportedGroup.name).toEqual(profileGroup.name)
  expect(reimportedGroup.profiles.length).toEqual(profileGroup.profiles.length)

  for (let i = 0; i < profileGroup.profiles.length; i++) {
    const profile = profileGroup.profiles[i]
    const reimported = reimportedGroup.profiles[i]

    expect(reimported.getTotalWeight()).toEqual(profile.getTotalWeight())
    expect(dumpProfile(reimported).stacks.join('\n')).toEqual(
      dumpProfile(profile).stacks.join('\n'),
    )
  }

  const reexported = exportProfileGroup(reimportedGroup)
  expect(exported).toEqual(reexported)
}
