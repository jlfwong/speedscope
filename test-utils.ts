import * as fs from 'fs'
import * as path from 'path'
import {Profile, CallTreeNode, Frame} from './profile'
import {importProfile} from './import'

interface DumpedProfile {
  stacks: string[]
  frames: Frame[]
}

export function dumpProfile(profile: Profile): any {
  const dump: DumpedProfile = {
    stacks: [],
    frames: [],
  }

  profile.forEachFrame(f => dump.frames.push(f))

  let lastValue = 0
  const curStack: (number | string)[] = []

  function maybeEmit(value: number) {
    if (lastValue != value) {
      dump.stacks.push(
        curStack.map(k => `${k}`).join(';') + ` ${profile.formatValue(value - lastValue)}`,
      )
      lastValue = value
    }
  }

  function openFrame(node: CallTreeNode, value: number) {
    maybeEmit(value)
    curStack.push(node.frame.name)
  }

  function closeFrame(node: CallTreeNode, value: number) {
    maybeEmit(value)
    curStack.pop()
  }

  profile.forEachCall(openFrame, closeFrame)

  return dump
}

export async function checkProfileSnapshot(filepath: string) {
  const input = fs.readFileSync(filepath, 'utf8')
  const profile = await importProfile(path.basename(filepath), input)
  if (profile) {
    expect(dumpProfile(profile)).toMatchSnapshot()
  } else {
    fail('Failed to extract profile')
  }
}
