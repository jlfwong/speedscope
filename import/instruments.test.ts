import * as fs from 'fs'
import {dumpProfile} from '../test-utils'
import {importFromInstrumentsDeepCopy} from './instruments'

describe('importFromInstrumentsDeepCopy', () => {
  test('time profile', () => {
    const input = fs.readFileSync(
      './sample/profiles/Instruments/7.3.1/simple-time-profile-deep-copy.txt',
      'utf8',
    )
    const profile = importFromInstrumentsDeepCopy(input)
    expect(dumpProfile(profile)).toMatchSnapshot()
  })

  test('allocations profile', () => {
    const input = fs.readFileSync(
      './sample/profiles/Instruments/7.3.1/random-allocations-deep-copy.txt',
      'utf8',
    )
    const profile = importFromInstrumentsDeepCopy(input)
    expect(dumpProfile(profile)).toMatchSnapshot()
  })
})
