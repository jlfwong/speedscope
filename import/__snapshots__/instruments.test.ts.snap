// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`importFromInstrumentsDeepCopy allocations profile 1`] = `
Object {
  "frames": Array [
    Frame {
      "col": undefined,
      "file": undefined,
      "key": ":start",
      "line": undefined,
      "name": "start",
      "selfWeight": 0,
      "totalWeight": 449839104,
    },
    Frame {
      "col": undefined,
      "file": undefined,
      "key": ":main",
      "line": undefined,
      "name": "main",
      "selfWeight": 1048576,
      "totalWeight": 449839104,
    },
    Frame {
      "col": undefined,
      "file": undefined,
      "key": ":delta()",
      "line": undefined,
      "name": "delta()",
      "selfWeight": 1048576,
      "totalWeight": 227540992,
    },
    Frame {
      "col": undefined,
      "file": undefined,
      "key": ":alpha()",
      "line": undefined,
      "name": "alpha()",
      "selfWeight": 0,
      "totalWeight": 150994944,
    },
    Frame {
      "col": undefined,
      "file": undefined,
      "key": ":leakMemory()",
      "line": undefined,
      "name": "leakMemory()",
      "selfWeight": 0,
      "totalWeight": 447741952,
    },
    Frame {
      "col": undefined,
      "file": undefined,
      "key": ":malloc",
      "line": undefined,
      "name": "malloc",
      "selfWeight": 0,
      "totalWeight": 447741952,
    },
    Frame {
      "col": undefined,
      "file": undefined,
      "key": ":malloc_zone_malloc",
      "line": undefined,
      "name": "malloc_zone_malloc",
      "selfWeight": 447741952,
      "totalWeight": 447741952,
    },
    Frame {
      "col": undefined,
      "file": undefined,
      "key": ":beta()",
      "line": undefined,
      "name": "beta()",
      "selfWeight": 0,
      "totalWeight": 147849216,
    },
    Frame {
      "col": undefined,
      "file": undefined,
      "key": ":gamma()",
      "line": undefined,
      "name": "gamma()",
      "selfWeight": 0,
      "totalWeight": 74448896,
    },
  ],
  "stacks": Array [
    "start;main;delta();alpha();leakMemory();malloc;malloc_zone_malloc 73.00 MB",
    "start;main;delta();beta();leakMemory();malloc;malloc_zone_malloc 72.00 MB",
    "start;main;delta();leakMemory();malloc;malloc_zone_malloc 71.00 MB",
    "start;main;delta() 1.00 MB",
    "start;main;alpha();leakMemory();malloc;malloc_zone_malloc 71.00 MB",
    "start;main;gamma();leakMemory();malloc;malloc_zone_malloc 71.00 MB",
    "start;main;beta();leakMemory();malloc;malloc_zone_malloc 69.00 MB",
    "start;main 1.00 MB",
  ],
}
`;

exports[`importFromInstrumentsDeepCopy time profile 1`] = `
Object {
  "frames": Array [
    Frame {
      "col": undefined,
      "file": undefined,
      "key": ":start",
      "line": undefined,
      "name": "start",
      "selfWeight": 0,
      "totalWeight": 4273,
    },
    Frame {
      "col": undefined,
      "file": undefined,
      "key": ":main",
      "line": undefined,
      "name": "main",
      "selfWeight": 0,
      "totalWeight": 4273,
    },
    Frame {
      "col": undefined,
      "file": undefined,
      "key": ":delta()",
      "line": undefined,
      "name": "delta()",
      "selfWeight": 693,
      "totalWeight": 2126,
    },
    Frame {
      "col": undefined,
      "file": undefined,
      "key": ":alpha()",
      "line": undefined,
      "name": "alpha()",
      "selfWeight": 1411,
      "totalWeight": 1438,
    },
    Frame {
      "col": undefined,
      "file": undefined,
      "key": ":leakMemory()",
      "line": undefined,
      "name": "leakMemory()",
      "selfWeight": 1,
      "totalWeight": 67,
    },
    Frame {
      "col": undefined,
      "file": undefined,
      "key": ":malloc",
      "line": undefined,
      "name": "malloc",
      "selfWeight": 3,
      "totalWeight": 66,
    },
    Frame {
      "col": undefined,
      "file": undefined,
      "key": ":malloc_zone_malloc",
      "line": undefined,
      "name": "malloc_zone_malloc",
      "selfWeight": 3,
      "totalWeight": 62,
    },
    Frame {
      "col": undefined,
      "file": undefined,
      "key": ":szone_malloc_should_clear",
      "line": undefined,
      "name": "szone_malloc_should_clear",
      "selfWeight": 21,
      "totalWeight": 39,
    },
    Frame {
      "col": undefined,
      "file": undefined,
      "key": ":small_malloc_from_free_list",
      "line": undefined,
      "name": "small_malloc_from_free_list",
      "selfWeight": 17,
      "totalWeight": 17,
    },
    Frame {
      "col": undefined,
      "file": undefined,
      "key": ":allocate_pages_securely",
      "line": undefined,
      "name": "allocate_pages_securely",
      "selfWeight": 0,
      "totalWeight": 2,
    },
    Frame {
      "col": undefined,
      "file": undefined,
      "key": ":mach_vm_map",
      "line": undefined,
      "name": "mach_vm_map",
      "selfWeight": 0,
      "totalWeight": 2,
    },
    Frame {
      "col": undefined,
      "file": undefined,
      "key": ":_kernelrpc_mach_vm_map_trap",
      "line": undefined,
      "name": "_kernelrpc_mach_vm_map_trap",
      "selfWeight": 2,
      "totalWeight": 2,
    },
    Frame {
      "col": undefined,
      "file": undefined,
      "key": ":_os_lock_spin_lock",
      "line": undefined,
      "name": "_os_lock_spin_lock",
      "selfWeight": 14,
      "totalWeight": 14,
    },
    Frame {
      "col": undefined,
      "file": undefined,
      "key": ":os_lock_lock",
      "line": undefined,
      "name": "os_lock_lock",
      "selfWeight": 3,
      "totalWeight": 3,
    },
    Frame {
      "col": undefined,
      "file": undefined,
      "key": ":beta()",
      "line": undefined,
      "name": "beta()",
      "selfWeight": 1406,
      "totalWeight": 1427,
    },
    Frame {
      "col": undefined,
      "file": undefined,
      "key": ":<Unknown Address>",
      "line": undefined,
      "name": "<Unknown Address>",
      "selfWeight": 0,
      "totalWeight": 2,
    },
    Frame {
      "col": undefined,
      "file": undefined,
      "key": ":szone_malloc",
      "line": undefined,
      "name": "szone_malloc",
      "selfWeight": 1,
      "totalWeight": 1,
    },
    Frame {
      "col": undefined,
      "file": undefined,
      "key": ":DYLD-STUB$$os_lock_unlock",
      "line": undefined,
      "name": "DYLD-STUB$$os_lock_unlock",
      "selfWeight": 1,
      "totalWeight": 1,
    },
    Frame {
      "col": undefined,
      "file": undefined,
      "key": ":gamma()",
      "line": undefined,
      "name": "gamma()",
      "selfWeight": 696,
      "totalWeight": 706,
    },
    Frame {
      "col": undefined,
      "file": undefined,
      "key": ":os_lock_unlock",
      "line": undefined,
      "name": "os_lock_unlock",
      "selfWeight": 1,
      "totalWeight": 1,
    },
  ],
  "stacks": Array [
    "start;main;delta();alpha();leakMemory();malloc;malloc_zone_malloc;szone_malloc_should_clear;small_malloc_from_free_list 4.00ms",
    "start;main;delta();alpha();leakMemory();malloc;malloc_zone_malloc;szone_malloc_should_clear;allocate_pages_securely;mach_vm_map;_kernelrpc_mach_vm_map_trap 1.00ms",
    "start;main;delta();alpha();leakMemory();malloc;malloc_zone_malloc;szone_malloc_should_clear 2.00ms",
    "start;main;delta();alpha();leakMemory();malloc;malloc_zone_malloc;_os_lock_spin_lock 5.00ms",
    "start;main;delta();alpha();leakMemory();malloc;malloc_zone_malloc;os_lock_lock 2.00ms",
    "start;main;delta();alpha();leakMemory();malloc;malloc_zone_malloc 1.00ms",
    "start;main;delta();alpha();leakMemory();malloc 1.00ms",
    "start;main;delta();alpha();leakMemory() 1.00ms",
    "start;main;delta();alpha() 696.00ms",
    "start;main;delta();beta();leakMemory();malloc;malloc_zone_malloc;szone_malloc_should_clear;small_malloc_from_free_list 1.00ms",
    "start;main;delta();beta();leakMemory();malloc;malloc_zone_malloc;szone_malloc_should_clear 6.00ms",
    "start;main;delta();beta();leakMemory();malloc;malloc_zone_malloc;_os_lock_spin_lock 2.00ms",
    "start;main;delta();beta() 702.00ms",
    "start;main;delta();leakMemory();malloc;malloc_zone_malloc;szone_malloc_should_clear;small_malloc_from_free_list 1.00ms",
    "start;main;delta();leakMemory();malloc;malloc_zone_malloc;szone_malloc_should_clear 4.00ms",
    "start;main;delta();leakMemory();malloc;malloc_zone_malloc;_os_lock_spin_lock 2.00ms",
    "start;main;delta();leakMemory();malloc;malloc_zone_malloc;os_lock_lock 1.00ms",
    "start;main;delta();leakMemory();malloc;<Unknown Address>;szone_malloc 1.00ms",
    "start;main;delta() 693.00ms",
    "start;main;alpha();leakMemory();malloc;malloc_zone_malloc;szone_malloc_should_clear;small_malloc_from_free_list 4.00ms",
    "start;main;alpha();leakMemory();malloc;malloc_zone_malloc;szone_malloc_should_clear 2.00ms",
    "start;main;alpha();leakMemory();malloc;malloc_zone_malloc;_os_lock_spin_lock 2.00ms",
    "start;main;alpha();leakMemory();malloc;malloc_zone_malloc 1.00ms",
    "start;main;alpha();leakMemory();malloc 1.00ms",
    "start;main;alpha() 715.00ms",
    "start;main;beta();leakMemory();malloc;malloc_zone_malloc;szone_malloc_should_clear;small_malloc_from_free_list 3.00ms",
    "start;main;beta();leakMemory();malloc;malloc_zone_malloc;szone_malloc_should_clear;allocate_pages_securely;mach_vm_map;_kernelrpc_mach_vm_map_trap 1.00ms",
    "start;main;beta();leakMemory();malloc;malloc_zone_malloc;szone_malloc_should_clear 5.00ms",
    "start;main;beta();leakMemory();malloc;malloc_zone_malloc;_os_lock_spin_lock 1.00ms",
    "start;main;beta();leakMemory();malloc;malloc_zone_malloc;DYLD-STUB$$os_lock_unlock 1.00ms",
    "start;main;beta();leakMemory();malloc 1.00ms",
    "start;main;beta() 704.00ms",
    "start;main;gamma();leakMemory();malloc;malloc_zone_malloc;szone_malloc_should_clear;small_malloc_from_free_list 3.00ms",
    "start;main;gamma();leakMemory();malloc;malloc_zone_malloc;szone_malloc_should_clear 2.00ms",
    "start;main;gamma();leakMemory();malloc;malloc_zone_malloc;_os_lock_spin_lock 2.00ms",
    "start;main;gamma();leakMemory();malloc;malloc_zone_malloc;os_lock_unlock 1.00ms",
    "start;main;gamma();leakMemory();malloc;malloc_zone_malloc;<Unknown Address>;small_malloc_from_free_list 1.00ms",
    "start;main;gamma();leakMemory();malloc;malloc_zone_malloc 1.00ms",
    "start;main;gamma() 696.00ms",
  ],
}
`;
